pkg/fake/iamapi.go

```diff
@@ -0,0 +1,150 @@
+			InstanceProfileName: input.InstanceProfileName,
+			Path:                input.Path,
+			Tags:                input.Tags,
+		}
+		s.InstanceProfiles[aws.StringValue(input.InstanceProfileName)] = instanceProfile
+		return &iam.CreateInstanceProfileOutput{InstanceProfile: instanceProfile}, nil
+	})
+}
+
+func (s *IAMAPI) DeleteInstanceProfileWithContext(_ context.Context, input *iam.DeleteInstanceProfileInput, _ ...request.Option) (*iam.DeleteInstanceProfileOutput, error) {
+	return s.DeleteInstanceProfileBehavior.Invoke(input, func(output *iam.DeleteInstanceProfileInput) (*iam.DeleteInstanceProfileOutput, error) {
+		s.Lock()
+		defer s.Unlock()
+
+		if i, ok := s.InstanceProfiles[aws.StringValue(input.InstanceProfileName)]; ok {
+			if len(i.Roles) > 0 {
+				return nil, awserr.New(iam.ErrCodeDeleteConflictException, "Cannot delete entity, must remove roles from instance profile first.", nil)
+			}
+			delete(s.InstanceProfiles, aws.StringValue(input.InstanceProfileName))
+			return &iam.DeleteInstanceProfileOutput{}, nil
+		}
+		return nil, awserr.New(iam.ErrCodeNoSuchEntityException, fmt.Sprintf("Instance Profile %s cannot be found", aws.StringValue(input.InstanceProfileName)), nil)
+	})
+}
+
+func (s *IAMAPI) AddRoleToInstanceProfileWithContext(_ context.Context, input *iam.AddRoleToInstanceProfileInput, _ ...request.Option) (*iam.AddRoleToInstanceProfileOutput, error) {
+	return s.AddRoleToInstanceProfileBehavior.Invoke(input, func(output *iam.AddRoleToInstanceProfileInput) (*iam.AddRoleToInstanceProfileOutput, error) {
+		s.Lock()
+		defer s.Unlock()
+
+		if i, ok := s.InstanceProfiles[aws.StringValue(input.InstanceProfileName)]; ok {
+			if len(i.Roles) > 0 {
+				return nil, awserr.New(iam.ErrCodeLimitExceededException, "Cannot exceed quota for InstanceSessionsPerInstanceProfile: 1", nil)
+			}
+			i.Roles = append(i.Roles, &iam.Role{RoleId: aws.String(RoleID()), RoleName: input.RoleName})
+			return nil, nil
+		}
+		return nil, awserr.New(iam.ErrCodeNoSuchEntityException, fmt.Sprintf("Instance Profile %s cannot be found", aws.StringValue(input.InstanceProfileName)), nil)
+	})
+}
+
+func (s *IAMAPI) RemoveRoleFromInstanceProfileWithContext(_ context.Context, input *iam.RemoveRoleFromInstanceProfileInput, _ ...request.Option) (*iam.RemoveRoleFromInstanceProfileOutput, error) {
+	return s.RemoveRoleFromInstanceProfileBehavior.Invoke(input, func(output *iam.RemoveRoleFromInstanceProfileInput) (*iam.RemoveRoleFromInstanceProfileOutput, error) {
+		s.Lock()
+		defer s.Unlock()
+
+		if i, ok := s.InstanceProfiles[aws.StringValue(input.InstanceProfileName)]; ok {
+			var newRoles []*iam.Role
+			for _, role := range i.Roles {
+				if aws.StringValue(role.RoleName) != aws.StringValue(input.RoleName) {
+					newRoles = append(newRoles, role)
+				}
+			}
+			if len(i.Roles) == len(newRoles) {
+				return nil, awserr.New(iam.ErrCodeNoSuchEntityException, fmt.Sprintf("The role with name %s cannot be found", aws.StringValue(input.RoleName)), nil)
```

Comment:
nit: Can probably shorten, since it's inferred that we're referring to the role by name
```suggestion
				return nil, awserr.New(iam.ErrCodeNoSuchEntityException, fmt.Sprintf("Role %s cannot be found", aws.StringValue(input.RoleName)), nil)
```

Reply:
I was mainly just copying the formatting of the error messages from the actual IAM apis. If you attempt these error paths on these APIs, this is the error you get.

Reply:
Ah if this is the same error as the APIs, sounds good to me 👍